### DISCUSSION TITLE
The cancel button on an Edit/Create Page should redirect to previous page

### DISCUSSION THREAD

**Discussion by flyingeek** [filamentphp/filament#4364] [Ideas]
When you are administrating the site and working on a filtered list of user in a certain order, if you edit a record (not in a modal) and you press Cancel you are redirected to a different page: the list with all states lost.

The only way to return to the previous page is to use the history back browser button instead of Cancel.

If possible it would be nice to have a getPreviousUrl method that we could also use to redirect after save.

**Comment by flyingeek** [comment]
For the cancel button here is what I do at the moment:

Override getCancelFormAction method where needed (ModelResource/Pages/Page)

```php
protected function getCancelFormAction(): Action
{
    return GoBackAction::make();
} 
```

```php
class GoBackAction
{
    static public function make($label): Action
    {
        return Action::make('go_back')
            ->label($label ?? __('filament::resources/pages/edit-record.form.actions.cancel.label'))
            ->extraAttributes(['onclick' => 'window.history.back()'])
            ->color('secondary');
    }
}
```

**Reply by hrvoj3e** [reply]
See https://github.com/filamentphp/filament/discussions/4079. I have a similar problem.

Also, keep in mind that action that has cancel could be called from modal so no navigation is desired in that case.

I think your solution is OK.

**Reply by flyingeek** [reply]
The trait solution I've just added is working perfectly for me, what about your usage ?

**Comment by flyingeek** [comment]
I have found a better way, may be a good coder could propose a PR:

Create the trait below then add to your EditPage or CreatePage the trait

```php
<?php

namespace App\Traits;

use Filament\Pages\Actions\Action;
use Filament\Resources\Pages\Page;

trait HasBackUrl
{
    public ?string $backUrl = null;

    public function mount($record=null): void
    {
        $this->backUrl = url()->previous();
        parent::mount($record);
    }

    protected function getRedirectUrl(): string  // <- return type is string to be compatible with EditRecord & CreateRecord
    {
        return $this->backUrl;
    }

    protected function getCancelFormAction(): Action
    {
        return parent::getCancelFormAction()
            ->url(fn(Page $livewire) => $livewire->backUrl ?? static::getResource()::getUrl());
    }
}
```

Then in your EditUser page (not the resource) add for exemple:

```php
use Filament\Resources\Pages\EditRecord;
use App\Traits\HasBackUrl;

class EditUser extends EditRecord
{
    use HasBackUrl;   // <- add this line
   //....
}
```

Note that with the trait added, you can still disallow save redirect this way:

```php
use Filament\Resources\Pages\EditRecord;
use App\Traits\HasBackUrl;

class EditUser extends EditRecord
{
    use HasBackUrl; 

    protected function getRedirectUrl(): string
    {
        return '';  // <- you have to return empty string not null
    }
   //....
}
```
